sling/pyapi/pyframe.cc

```diff
@@ -0,0 +1,365 @@
+  if (role.IsNil()) Py_RETURN_NONE;
+
+  // Look up (first) value for role.
+  Handle value = frame()->get(role);
+  return pystore->PyValue(value);
+}
+
+int PyFrame::Assign(PyObject *key, PyObject *v) {
+  // Check that frame is writable.
+  if (!Writable()) return -1;
+
+  // Look up role.
+  GCLock lock(pystore->store);
+  Handle role = pystore->RoleValue(key);
+  if (role.IsError()) return -1;
+
+  // Get new frame role value.
+  Handle value = pystore->Value(v);
+  if (value.IsError()) return -1;
+
+  // Check role.
+  if (role.IsNil()) {
+    PyErr_SetString(PyExc_IndexError, "Role not defined");
+    return -1;
+  };
+  if (role == Handle::id()) {
+    PyErr_SetString(PyExc_IndexError, "Frame id cannot be changed");
+    return -1;
+  };
+
+  // Set frame slot value.
+  pystore->store->Set(handle(), role, value);
+
+  return 0;
+}
+
+int PyFrame::Contains(PyObject *key) {
+  // Look up role.
+  Handle role = pystore->RoleValue(key, true);
+  if (role.IsError()) return -1;
+
+  // Check if frame has slot with role.
+  if (role.IsNil()) return 0;
+  return frame()->has(role);
+}
+
+PyObject *PyFrame::GetAttr(PyObject *key) {
+  // Get attribute name.
+  char *name = PyString_AsString(key);
+  if (name == nullptr) return nullptr;
+
+  // Resolve methods.
+  PyObject *method = Py_FindMethod(methods, AsObject(), name);
+  if (method != nullptr) return method;
+  PyErr_Clear();
+
+  // Lookup role.
+  Handle role = pystore->store->LookupExisting(name);
+  if (role.IsNil()) Py_RETURN_NONE;
+
+  // Get role value for frame.
+  Handle value = frame()->get(role);
+  return pystore->PyValue(value);
+}
+
+int PyFrame::SetAttr(PyObject *key, PyObject *v) {
+  // Check that frame is writable.
+  if (!Writable()) return -1;
+
+  // Get role name.
+  char *name = PyString_AsString(key);
+  if (name == nullptr) return -1;
+
+  // Lookup role.
+  GCLock lock(pystore->store);
+  Handle role = pystore->store->Lookup(name);
+  if (role.IsNil()) {
+    PyErr_SetString(PyExc_IndexError, "Role not defined");
+    return -1;
+  };
+  if (role == Handle::id()) {
+    PyErr_SetString(PyExc_IndexError, "Frame id cannot be changed");
+    return -1;
+  };
+
+  // Get role value.
+  Handle value = pystore->Value(v);
+  if (value.IsError())  return -1;
+
+  // Set role value for frame.
+  pystore->store->Set(handle(), role, value);
+
+  return 0;
+}
+
+PyObject *PyFrame::Append(PyObject *args) {
+  // Check that frame is writable.
+  if (!Writable()) return nullptr;
+
+  // Get name and value arguments.
+  PyObject *pyname;
+  PyObject *pyvalue;
+  if (!PyArg_ParseTuple(args, "OO", &pyname, &pyvalue)) return nullptr;
+
+  // Get role.
+  GCLock lock(pystore->store);
+  Handle role = pystore->RoleValue(pyname);
+  if (role.IsError()) return nullptr;
+
+  // Get value.
+  Handle value = pystore->Value(pyvalue);
+  if (value.IsError()) return nullptr;
+
+  // Set role value for frame.
+  pystore->store->Set(handle(), role, value);
```

Comment:
In case the frame has an existing 'role' slot, Store::Set() overwrites the slot value instead of appending.
Perhaps you meant to call Store::Add() instead?

Reply:
Yes, you're right. It should be Add and not Set. Fixed.